presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,129 @@
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            if (memo.containsKey(joinGraph)) {
+                return memo.get(joinGraph);
+            }
+            else {
+                PriorityQueue<JoinNode> joinOrders = planPriorityQueue(symbolAllocator, lookup, session, costComparator);
+                // TODO: eliminate cross joins
+                for (Set<Integer> partitioning : generatePartitions(joinGraph.getSources().size())) {
+                    joinOrders.add(createJoinNodeTree(joinGraph, partitioning, idAllocator, symbolAllocator, lookup, session));
+                }
+                checkState(joinOrders.size() >= 1, "joinOrders cannot be empty");
+                JoinNode node = joinOrders.poll();
+                memo.put(joinGraph, node);
+                return node;
+            }
+        }
+
+        private JoinNode createJoinNodeTree(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            JoinNode placeholderJoin = createJoinAccordingToPartitioning(joinGraph, partitioning, idAllocator);
+            PlanNode left = lookup.resolve(placeholderJoin.getLeft());
+            PlanNode right = lookup.resolve(placeholderJoin.getRight());
+            if (left instanceof JoinGraphNode) {
+                left = chooseJoinOrder((JoinGraphNode) left, idAllocator, symbolAllocator, lookup, session);
+            }
+            if (right instanceof JoinGraphNode) {
+                right = chooseJoinOrder((JoinGraphNode) right, idAllocator, symbolAllocator, lookup, session);
+            }
+
+            JoinNode joinNode = new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    placeholderJoin.getCriteria(),
+                    placeholderJoin.getOutputSymbols(),
+                    placeholderJoin.getFilter(),
+                    placeholderJoin.getLeftHashSymbol(),
+                    placeholderJoin.getRightHashSymbol(),
+                    Optional.empty());
+            return chooseJoinNodeProperties(joinNode, symbolAllocator, lookup, session);
+        }
+
+        private JoinNode chooseJoinNodeProperties(JoinNode joinNode, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            PriorityQueue<JoinNode> possibleJoinNodes = planPriorityQueue(symbolAllocator, lookup, session, costComparator);
+            String joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPARTITIONED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
+                JoinNode repartitionedJoin = joinNode.withDistributionType(PARTITIONED);
```

Comment:
partitioned should probably be renamed to repartitioned.  I think we got rid of distributed as the name for a  repartitioned join.